src/components/TodoContext.tsx

```diff
@@ -0,0 +1,219 @@
+  handleClearCompleted: () => {},
+  tmpTodo: null,
+  handleTmpTodo: () => {},
+  isLoadingAll: false,
+  handleLoadingAll: () => {},
+});
+
+export type Props = {
+  children: React.ReactNode;
+};
+
+export enum ActionNames {
+  Load = 'LOAD',
+  Add = 'ADD',
+  Delete = 'DELETE',
+  Update = 'UPDATE',
+  Completed = 'TOGGLE_COMPLETED',
+  ToggleAllCompleted = 'TOGGLE_ALL_COMPLETED',
+  ClearCompleted = 'CLEAR_COMPLETED',
+}
+
+export type Action =
+  | { type: ActionNames.Load; payload: Todo[] }
+  | { type: ActionNames.Add; payload: Todo }
+  | { type: ActionNames.Delete; payload: number }
+  | { type: ActionNames.Update; payload: { id: number; title: string } }
+  | { type: ActionNames.Completed; payload: { id: number; completed: boolean } }
+  | { type: ActionNames.ToggleAllCompleted }
+  | { type: ActionNames.ClearCompleted };
+
+export const errors = {
+  LoadTodos: 'Unable to load todos',
+  EmptyTitle: 'Title should not be empty',
+  AddTodo: 'Unable to add a todo',
+  DeleteTodo: 'Unable to delete a todo',
+  UpdateTodo: 'Unable to update a todo',
+};
+
+function reducer(state: Todo[], action: Action) {
+  switch (action.type) {
+    case ActionNames.Load:
+      return action.payload;
+
+    case ActionNames.Add:
+      return [...state, action.payload];
+
+    case ActionNames.Delete:
+      return state.filter(todo => todo.id !== action.payload);
+
+    case ActionNames.Update:
+      return state.map((todo: Todo) => {
+        if (todo.id === action.payload.id) {
+          return {
+            ...todo,
+            title: action.payload.title.trim(),
+          };
+        }
+
+        return todo;
+      });
+
+    case ActionNames.Completed:
+      return state.map(todo => {
+        if (todo.id === action.payload.id) {
+          return {
+            ...todo,
+            completed: !todo.completed,
+          };
+        }
+
+        return todo;
+      });
+
+    case ActionNames.ToggleAllCompleted:
+      const completed = state.some(todo => todo.completed === false);
+
+      return state.map(todo => {
+        return {
+          ...todo,
+          completed: completed,
+        };
+      });
+
+    case ActionNames.ClearCompleted:
+      return state.filter(todo => !todo.completed);
+
+    default:
+      return state;
+  }
+}
+
+function filter(todos: Todo[], type: FilterAction) {
+  switch (type) {
+    case FilterBy.Active:
+      return todos.filter(todo => !todo.completed);
+
+    case FilterBy.Completed:
+      return todos.filter(todo => todo.completed);
+
+    case FilterBy.All:
+    default:
+      return todos;
+  }
+}
+
+export const TodoProvider: React.FC<Props> = ({ children }) => {
+  const [todos, dispatch] = useReducer(reducer, []);
+  const [tmpTodo, setTmpTodo] = useState<Todo | null>(null);
+  const [filteredBy, setFilteredBy] = useState<FilterBy>(FilterBy.All);
+  const [error, setError] = useState<string | null>(null);
+  const [deleteCandidates, setDeleteCandidates] = useState<number[]>([]);
```

Comment:
you don't need `deleteCandidates` state. You can just remove todos from `todos` state based on `todo.completed`

Reply:
Dind't get your point, I have this state not only for deleting from server I also use it for loading when I delete all completed. Could you be more specific.

Reply:
To show loader you can have just another loadingIds state and store there an array of todoIds that are deleting

Reply:
@SanyaBratashchuk why would i need another one if I alredy have it?